tests/MongoDB.Driver.Tests/Specifications/unified-test-format/UnifiedValueMatcher.cs

```diff
@@ -0,0 +1,207 @@
+        {
+            _entityMap = entityMap;
+        }
+
+        public void AssertValuesMatch(BsonValue expected, BsonValue actual)
+        {
+            AssertValuesMatch(expected, actual, isRoot: true);
+        }
+
+        private void AssertValuesMatch(BsonValue expected, BsonValue actual, bool isRoot)
+        {
+            if (expected.IsBsonDocument &&
+                expected.AsBsonDocument.ElementCount == 1 &&
+                expected.AsBsonDocument.GetElement(0).Name.StartsWith("$$"))
+            {
+                var expectedDocument = expected.AsBsonDocument; // TODO: Recheck if non document allowed
+                switch (expectedDocument.GetElement(0).Name)
+                {
+                    case "$$unsetOrMatches":
+                        // TODO: recheck
+                        if (actual == null)
+                        {
+                            return;
+                        }
+                        expected = expectedDocument[0];
+                        break;
+                    case "$$type":
+                        AssertExpectedType(actual, expectedDocument[0]);
+                        return;
+                    case "$$matchesHexBytes":
+                        // TODO: recheck
+                        expected = expectedDocument[0];
+                        break;
+                    default:
+                        throw new NotSupportedException($"Special operator not supported: '{expectedDocument.GetElement(0).Name}'");
+                }
+            }
+
+            if (expected.IsBsonDocument)
+            {
+                actual.BsonType.Should().Be(BsonType.Document);
+
+                var expectedDocument = expected.AsBsonDocument;
+                var actualDocument = actual.AsBsonDocument;
+
+                foreach (var expectedElement in expectedDocument)
+                {
+                    BsonValue expectedItem = expectedElement.Value;
+                    // TODO: consider using var t = expectedElement.Value is BsonDocument expectedDocument1;
+                    if (expectedElement.Value.IsBsonDocument &&
+                        expectedElement.Value.AsBsonDocument.ElementCount == 1 &&
+                        expectedElement.Value.AsBsonDocument.GetElement(0).Name.StartsWith("$$"))
+                    {
+                        var specialOperator = expectedElement.Value.AsBsonDocument;
+                        switch (specialOperator.GetElement(0).Name) // TODO: Recheck the switch order
+                        {
+                            case "$$exists":
+                                actualDocument.Contains(expectedElement.Name).Should().Be(specialOperator[0].AsBoolean); // TODO: Recheck this actually works
+                                continue;
+                            case "$$type":
+                                AssertExpectedType(actual, specialOperator[0]);
+                                continue;
```

Comment:
Deeply nested `continue` statements are hard to comprehend because it's not always easy to find where execution will continue after this line...